### PR TITLE
Feature/auth flows

### PR DESCRIPTION
**Pull request by lykmapipo**
- Implement Initial Auth logics
- Save current signin party(or user) in shared preferences
- Add base SecureCompact for forcing auth logic on extending activities

### CODE REVIEW

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+public class SigninActivity extends AppCompatActivity {
+
+    public static final String TAG = SigninActivity.class.getSimpleName();
+
+    //view references
+    private TextInputLayout emailInputLayout;
+    private TextInputLayout passwordInputLayout;
+    private TextInputEditText emailTextInput;
+    private TextInputEditText passwordTextInput;
+    private Button signInButton;
```
Comment (krtonga) [inline majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:55]: I've been, so far, using the initials of the type, followed by its purpose so: `til_email` as id and `mTilEmail` or `tilEmail` as variable name. These names are clear and so I'm not too worried, but at some point we should discuss and decide for consistently.

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+    //rxbindings
+    private CompositeDisposable compositeDisposable;
+    private Observable<CharSequence> emailObservable;
+    private Observable<CharSequence> passwordObservable;
+    private Observable<Object> signInButtonObservable;
```
Comment (krtonga) [inline majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:61]: I'm also used to seeing `m` prefix for member variables.

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+     * init subscription
+     */
+    private void subscribe() {
+
+        //observe email text input changes
+        Disposable emailDisposable = emailObservable
+                .doOnNext(new Consumer<CharSequence>() {
+                    @Override
+                    public void accept(CharSequence charSequence) throws Exception {
+                        hideEmailValidationError();
```
Comment (krtonga) [inline majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:111]: This causes the error to "jump" with each character insert. Why are you clearing here as well as after verification?

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+                            @Override
+                            public void accept(Boolean areValidInputs) throws Exception {
+                                if (areValidInputs) {
+                                    enableSignInButton(); // enable signin button
+                                } else {
+                                    disableSignInButton(); // disable signin button
+                                }
+                            }
+                        });
+        compositeDisposable.add(signInButtonDisposable);
```
Comment (krtonga) [inline majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:184]: Cool.

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+                    emailInputLayout.getChildAt(1).setVisibility(View.GONE);
+                }
+                emailInputLayout.setError(null);
+                break;
+            case 2: // for show error
+                if (emailInputLayout.getChildCount() == 2) {
+                    emailInputLayout.getChildAt(1).setVisibility(View.VISIBLE);
+                }
+                emailInputLayout.setError(getString(R.string.auth_error_invalid_email));
+                break;
```
Comment (krtonga) [inline majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:290]: What is the need of changing the visibility of the Error to `GONE`, especially as this causes the fields to jump their location. Is this the intended behavior?

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+                    passwordInputLayout.getChildAt(1).setVisibility(View.GONE);
+                }
+                passwordInputLayout.setError(null);
+                break;
+            case 2: // for show error
+                if (passwordInputLayout.getChildCount() == 2) {
+                    passwordInputLayout.getChildAt(1).setVisibility(View.VISIBLE);
+                }
+                passwordInputLayout.setError(getString(R.string.auth_error_invalid_password));
+                break;
```
Comment (krtonga) [inline majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:321]: same question as above...

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+    protected void showMessage(int resId) {
+        String message = getString(resId);
+        showMessage(message);
+    }
+
+    /**
+     * Hide keyboard
+     * TODO use KeyboardUtils
+     */
+    public void hideKeyboard() {
```
Comment (krtonga) [inline majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:415]: This method can be found in `KeyboardUtils`

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java (hunk 1 above)
Comment (lykmapipo) [reply majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:55]: Sure...we should pick a convection here...on `layouts` it make sense to have layout initials(ti, til etc) `prefixing ids` and in codes normally i prefer to favour code readability e.g pronouncing `tilEmail` is different that `emailTextInputLayout` unless your write down the rules. But sure we can resonate on this.

File: majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java (hunk 2 above)
Comment (lykmapipo) [reply majifix311/src/main/java/com/example/majifix311/ui/auth/SigninActivity.java:61]: Yes it used mostly in Google Code Guidelines for Android. My proplem comes when you have `protected` and `private` member...any differentiation?. Also i oppose it due to code readability. It easy to pronounce `emailObservable` than `mEmailObservable`. Same applies to other classes. 

But am open to discussions, the only problem am having is I normally want everyone to read my code and understand them.

Lets pass over [https://github.com/ribot/android-guidelines/blob/master/project_and_code_guidelines.md](https://github.com/ribot/android-guidelines/blob/master/project_and_code_guidelines.md),
 [https://google.github.io/styleguide/javaguide.html](https://google.github.io/styleguide/javaguide.html),

**Pull request merged**